File: rad/engine/02_catalog/store/mode.go
package store

import (
	"context"
	"fmt"

	kv "github.com/Southclaws/rad/rad/engine/01_kv"
	"github.com/Southclaws/rad/rad/engine/02_catalog/model"
)

const modeKey = "/rad/catalog/meta/mode"

func ReadMode(ctx context.Context, view kv.KV) (model.Mode, error) {
	mode, ok, err := ReadStoredMode(ctx, view)
	if err != nil {
		return "", err
	}
	if !ok {
		return model.ModeDirect, nil
	}
	return mode, nil
}

func ReadStoredMode(ctx context.Context, view kv.KV) (model.Mode, bool, error) {
	raw, ok, err := view.Get(ctx, []byte(modeKey))
	if err != nil {
		return "", false, err
	}
	if !ok {
		return "", false, nil
	}
	mode, err := model.ParseMode(string(raw))
	if err != nil {
		return "", false, fmt.Errorf("catalog: corrupt stored mode %q", raw)
	}
	return mode, true, nil
}

func SetMode(ctx context.Context, view kv.KV, mode model.Mode) error {
	return view.Put(ctx, []byte(modeKey), []byte(mode))
}

File: src/engine/02_catalog/store/mode.rs
use std::str::FromStr;

use bytes::Bytes;

use crate::engine::catalog::model::Mode;
use crate::engine::catalog::{Error, ErrorKind, Result};
use crate::engine::kv::KvView;

use super::map_kv;

const MODE_KEY: &[u8] = b"/rad/catalog/meta/mode";

pub async fn read_mode<V: KvView + ?Sized>(view: &mut V) -> Result<Mode> {
    Ok(read_stored_mode(view).await?.unwrap_or(Mode::Direct))
}

pub async fn read_stored_mode<V: KvView + ?Sized>(view: &mut V) -> Result<Option<Mode>> {
    let Some(raw) = view.get(MODE_KEY).await.map_err(map_kv)? else {
        return Ok(None);
    };
    let value = std::str::from_utf8(&raw).map_err(|error| {
        Error::source(
            ErrorKind::CatalogCorrupt,
            format!("catalog: corrupt stored mode {raw:?}"),
            error,
        )
    })?;
    Mode::from_str(value).map(Some).map_err(|_| {
        Error::message(
            ErrorKind::CatalogCorrupt,
            format!("catalog: corrupt stored mode {raw:?}"),
        )
    })
}

pub async fn set_mode<V: KvView + ?Sized>(view: &mut V, mode: Mode) -> Result<()> {
    let value = match mode {
        Mode::Direct => "direct",
        Mode::Schema => "schema",
    };
    view.put(
        Bytes::from_static(MODE_KEY),
        Bytes::from_static(value.as_bytes()),
    )
    .await
    .map_err(map_kv)
}

#[cfg(test)]
mod tests {
    use crate::engine::kv::TransactionalKv;
    use crate::engine::kv::slatedb;

    use super::*;

    #[tokio::test]
    async fn missing_mode_defaults_to_direct_and_stored_mode_round_trips() {
        let mut database = slatedb::Store::memory("catalog-mode").await.unwrap();
        assert_eq!(read_mode(&mut database).await.unwrap(), Mode::Direct);
        assert_eq!(read_stored_mode(&mut database).await.unwrap(), None);
        set_mode(&mut database, Mode::Schema).await.unwrap();
        assert_eq!(read_mode(&mut database).await.unwrap(), Mode::Schema);
        database.close().await.unwrap();
    }
}
